[8.0] [ADD] pos order to sale order (w/o picking)

legalsylvain opened the pull request:
This PR add a new module pos_order_to_sale order that allow to create sale order based on draft pos order.
3 options are available : **create Draft Order**, **Create Confirmed Order**, or **Create delivered Picking**.

* Configuration is available on pos.config : 
![pos_config_form](https://user-images.githubusercontent.com/3407482/34638769-1b10ca32-f2d3-11e7-9100-25deac408cb7.png)

* New button is available on PoS : 
![pos_create_picking_option_1](https://user-images.githubusercontent.com/3407482/28163239-9afe8444-67ca-11e7-8cd2-1a627bebaa6f.png)

this PR is an alternative to another WIP module, made by Akretion team 'pos_sale_order'. (but the other module, completely remove pos.order feature). In that module, user can continue to create pos.order, and create sale.order (and delivered picking), if desired.
Ref OCA PR : https://github.com/OCA/pos/pull/35 and other work here : https://github.com/stefanrijnhart/pos/tree/8.0-pos-sale-order.

CC : @sebastienbeau, @chafique-delli, @houssine78, @hparfr, @StefanRijnhart.
@grap/developers


More detailled feature available here : 
https://github.com/grap/pos/blob/8.0_ADD_pos_order_to_sale_picking/pos_order_to_sale_order/README.rst

legalsylvain committed:
[ADD] new module pos_order_to_sale_order

elicoidal commented on pos_order_to_sale_order/README.rst:
```diff
@@ -0,0 +1,130 @@
+=======================
+
+
+This module extends the functionality of point of sale to allow you create
```
to allow sales order creation from the Point of Sales.

elicoidal commented on pos_order_to_sale_order/README.rst:
```diff
@@ -0,0 +1,130 @@
+This module extends the functionality of point of sale to allow you create
+Sale Orders from the Point of sale.
+
+It adds a new button, below the current bill on the point of sale front end UI,
```
In the POS UI, a button has been added to create the sales order and discard the current POS order

elicoidal commented on pos_order_to_sale_order/README.rst:
```diff
@@ -0,0 +1,130 @@
+that allows you to create a Sale order, based on the current PoS Order lines
+and then to discard the current PoS Order.
+
+This module is usefull if you have some customers that come every day in your
```
I think it is a bit limited: this is not the only case I know. Another potential usage is the possibility to have a tablet in a fair and take orders with a very simple interface
Maybe add "For example"

elicoidal commented on pos_order_to_sale_order/models/pos_config.py:
```diff
@@ -0,0 +1,35 @@
+    iface_create_sale_order_action = fields.Selection(
+        string='Sale Order Actions', default='delivered_picking',
+        required=True, selection=CREATE_SALE_ORDER_ACTION_SELECTION,
+        help="Choose wich operation will be done after creating a Sale Order"
```
Well... The selection is dynamic but the help message is not :wink: (none blocking)

legalsylvain replied:
You mean, the warning message of the PoS front end ?

elicoidal commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,137 @@
+                }
+                self.pos.pos_widget.screen_selector.show_popup('confirm', {
+                    message: _t('Create Sale Order and discard the current PoS Order ?'),
+                    comment: _t("This operation will permanently destroy the current PoS Order and create a Sale Order, based on the current order lines. You'll have to create later an invoice.\n Note if you had manually changed unit prices for some products, this changes will not been taken into account in the sale order, and should be done manually on the invoice again."),
```
s/destroy/discard

elicoidal reviewed (changes requested):
:+1: good contribution!
some details

legalsylvain commented:
Hi, 

@elicoidal : Thanks a lot for your review. I'll fix the description in the next days.

I just talked with @sebastienbeau, about this module, and we was imagining having the possibility to create an invoice too. (based on the delivered products), and even to create a paid invoice (with account payment) that mark the invoice as paid.

could you be interested by such features ?

hparfr commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,137 @@
+                    return;
+                }
+                // Check Customer
+                if (current_order.get('client') ===  null){
```
`if (!current_order.get('client'))`

Are you sure get() returns null and not Undefined ?

legalsylvain replied:
Well, it works yes. Do you want I test with Undefined, because it's a best practice ?

hparfr replied:
```js 
if (!current_order.get('client'))
```
Less strict, more conscise

hparfr commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,137 @@
+                    comment: _t("This operation will permanently destroy the current PoS Order and create a Sale Order, based on the current order lines. You'll have to create later an invoice.\n Note if you had manually changed unit prices for some products, this changes will not been taken into account in the sale order, and should be done manually on the invoice again."),
+                    confirm: function(){
+                        var SaleOrderModel = new instance.web.Model('sale.order');
+                        SaleOrderModel.call('create_order_from_pos', [self.prepare_create_sale_order(current_order)]
```
You may `return SaleOrderModel.call(...` for somewhere else.

legalsylvain replied:
Not sure to understand. renderElement doesn't return a deferred (in the native Odoo). So it will not be interpreted and managed ?

hparfr commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,137 @@
+                        ).then(function (result) {
+                            self.hook_create_sale_order_success(result);
+                        }).fail(function (error, event){
+                            self.hook_create_sale_order_error(error, event);
```
and re-throw a failed promise

legalsylvain replied:
same remark here.

hparfr replied:
It was a suggestion of improvement.
It's ok as is.

hparfr reviewed (approved):
Few remarks which are not blocking.
Thanks for this great work @legalsylvain 
👍

legalsylvain reviewed (commented):
Thanks for you review @hparfr . Do you have any thought about my proposal to include paid invoices ?

hparfr commented on pos_order_to_sale_order/README.rst:
```diff
@@ -0,0 +1,130 @@
+
+This module is usefull if you have some customers that come every day in your
+shop, but want to have a unique invoice at the end of the month. With that
+module, you can create a sale order adn deliver products every time to keep
```
:s/adn/and/

elicoidal commented:
> could you be interested by such features ?

Personally not (at least not for my customers) but having a quick entry for sales order and invoices looks like a very good idea for Very Small Businesses

tjgurwara99 commented:
This module doesn't take into account of the taxes in sales order. Has anyone been able to get a solution for it! I am new in Odoo community and will really appreciate the help.

legalsylvain committed:
[REF] pos_order_to_sale_order : allow to display multiple buttons ; [FIX] comments and descriptions

legalsylvain commented:
Hi all. Sorry for the late. 
@elicoidal : I updated the description according to your remarks. Could you update your review ?

@hparfr : the same, except two JS remarks. (promise, promise, ...) need help !

Note : I changed the field selection on pos.config by three checkbox. So : 
- it allows to have custom comment in the confirmation popup
- it allows to display many buttons in the same time, if desired

elicoidal commented on pos_order_to_sale_order/models/pos_config.py:
```diff
@@ -0,0 +1,27 @@
+class PosConfig(models.Model):
+    _inherit = 'pos.config'
+
+    iface_create_draft_sale_order = fields.Boolean(
```
Just a remark: it could have been interesting to have a checkbox "Create Sales Order" and a selection box with all possible states (you give 3 but multiple others could be imaginable).

elicoidal replied:
(Saw your previous message later :tongue: anyway both usability options have pros and cons :smile: )

elicoidal commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,173 @@
+            this.sale_order_state = options.sale_order_state;
+            if (this.sale_order_state == 'draft') {
+                this.display_text = _t("Create Draft Order");
+                this.confirmation_message = _t('Create Draft Sale Order and discard the current PoS Order ?');
```
remove space before '?' (multiple)

elicoidal commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -0,0 +1,173 @@
+            if (this.sale_order_state == 'draft') {
+                this.display_text = _t("Create Draft Order");
+                this.confirmation_message = _t('Create Draft Sale Order and discard the current PoS Order ?');
+                this.confirmation_comment = _t("This operation will permanently discard the current PoS Order and create a draft Sale Order, based on the current order lines. Note if you had manually changed unit prices for some products, this changes will not been taken into account in the sale order.");
```
had changed->have changed
(shouldnot this limitation be highlighted in the README BTW)

legalsylvain replied:
It is written in the readme.

elicoidal reviewed (approved):
some details.
LGTM

hparfr commented on pos_order_to_sale_order/static/src/js/pos_order_to_sale_order.js:
```diff
@@ -71,7 +82,10 @@ openerp.pos_order_to_sale_order = function(instance, local) {
         prepare_create_sale_order: function(order) {
-            return order.export_as_JSON();
+            var res = order.export_as_JSON();
+            console.log(res);
```
console.log

legalsylvain committed:
[FIX] typo

legalsylvain committed:
[FIX] typo

legalsylvain committed:
[FIX] remove console.log

legalsylvain commented:
@elicoidal : typo fixed. Now ok ?
thanks.

elicoidal reviewed (approved):
LGTM
